File: benchmarks/v3/scenarios/05-java-saga/without-mcp/src/main/java/com/example/saga/application/saga/SagaOrchestrator.java
package com.example.saga.application.saga;

import java.util.List;

/**
 * Interface for saga orchestrator that coordinates the execution of saga steps.
 */
public interface SagaOrchestrator {

    /**
     * Executes the saga with all registered steps.
     *
     * @param context the initial saga context
     * @return the result of the saga execution
     */
    SagaExecutionResult execute(SagaContext context);

    /**
     * Gets the list of registered steps.
     */
    List<SagaStep<?>> getSteps();
}

File: benchmarks/v3/scenarios/05-java-saga/with-mcp/domain/SagaOrchestrator.ts
import { SagaContext } from './SagaContext';
import { SagaStep } from './SagaStep';

/**
 * Result of a saga execution
 */
export interface SagaExecutionResult {
  readonly success: boolean;
  readonly context: SagaContext;
  readonly completedSteps: string[];
  readonly failedStep?: string;
  readonly error?: Error;
  readonly compensationResults?: CompensationSummary;
}

export interface CompensationSummary {
  readonly triggered: boolean;
  readonly completedCompensations: string[];
  readonly failedCompensations: string[];
}

/**
 * Interface for saga orchestrator
 */
export interface SagaOrchestrator {
  /**
   * Execute the saga with all registered steps
   * @param context Initial saga context
   * @returns Result of the saga execution
   */
  execute(context: SagaContext): Promise<SagaExecutionResult>;

  /**
   * Get the list of registered steps
   */
  getSteps(): ReadonlyArray<SagaStep>;
}
